src/internal/pfsdb/common.go

```diff
@@ -0,0 +1,199 @@
+package pfsdb
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/jmoiron/sqlx"
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pachsql"
+	"github.com/pachyderm/pachyderm/v2/src/internal/stream"
+)
+
+type (
+	SortOrder      int
+	FilterOperator string
+)
+
+const (
+	SortNone SortOrder = iota
+	SortAscend
+	SortDescend
+)
+
+const (
+	// Note that we use `?` as the placeholder for bindvar, but postgres uses $1, $2, etc.
+	// We solve this by rebinding the query before executing it using sqlx.Rebind.
+	Equal              FilterOperator = "%s = ?"
+	GreaterThan        FilterOperator = "%s > ?"
+	GreaterThanOrEqual FilterOperator = "%s >= ?"
+	LessThan           FilterOperator = "%s < ?"
+	LessThanOrEqual    FilterOperator = "%s <= ?"
+	NotEqual           FilterOperator = "%s <> ?"
+	ValueIn            FilterOperator = "%s in (?)"
+)
+
+type (
+	ModelType interface{ Repo | Commit | Branch }
+	FieldType interface{ CommitField | BranchField }
+	PairType  interface{ CommitPair | BranchPair }
+)
+
+type Filter[T FieldType] struct {
+	Field  T
+	Value  any
+	values []any
+	Op     FilterOperator
+}
+
+func (f *Filter[T]) QueryString() (formatted string, err error) {
+	formatted = fmt.Sprintf(string(f.Op), f.Field)
+	if f.Op == ValueIn {
+		formatted, f.values, err = sqlx.In(formatted, f.Value) // Thank god for this!
+		if err != nil {
+			return "", errors.Wrapf(err, "expanding %s", formatted)
+		}
+	}
+	return
+}
+
+type AndFilters[T FieldType] []*Filter[T]
```

Comment:
Why is AndFilter a special type of filter?  And is a binary operator like `=`, `<`, etc. above.  It gets weird when the query builder accepts an AndFilter but no other type of filter.  Seems like not the right design.

Reply:
So the idea behind this was that currently we only support `AND`ing expressions in our WHERE clauses. This is just a way to indicate to the user that all the filters being passed in are `AND`ed together.